pkg/services/searchV2/index.go

```diff
@@ -0,0 +1,399 @@
+package searchV2
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/searchV2/extract"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+	"github.com/grafana/grafana/pkg/services/store"
+)
+
+type dashboardLoader interface {
+	LoadDashboards(ctx context.Context, orgID int64, dashboardUID string) ([]dashboard, error)
+}
+
+// TODO: migrate to a more generic Watch interface to listen for changes.
+type eventStore interface {
+	GetLastEvent(ctx context.Context) (*store.EntityEvent, error)
+	GetAllEventsAfter(ctx context.Context, id int64) ([]*store.EntityEvent, error)
+}
+
+type dashboardIndex struct {
+	mu         sync.RWMutex
+	loader     dashboardLoader
+	dashboards map[int64][]dashboard
+	eventStore eventStore
+	logger     log.Logger
+}
+
+type dashboard struct {
+	id       int64
+	uid      string
+	isFolder bool
+	folderID int64
+	slug     string
+	created  time.Time
+	updated  time.Time
+	info     *extract.DashboardInfo
+}
+
+func newDashboardIndex(dashLoader dashboardLoader, evStore eventStore) *dashboardIndex {
+	return &dashboardIndex{
+		loader:     dashLoader,
+		eventStore: evStore,
+		dashboards: map[int64][]dashboard{},
+		logger:     log.New("dashboardIndex"),
+	}
+}
+
+func (i *dashboardIndex) run(ctx context.Context) error {
+	fullReIndexTicker := time.NewTicker(5 * time.Minute)
+	defer fullReIndexTicker.Stop()
+
+	partialUpdateTicker := time.NewTicker(5 * time.Second)
+	defer partialUpdateTicker.Stop()
+
+	var lastEventID int64
+	lastEvent, err := i.eventStore.GetLastEvent(context.Background())
+	if err != nil {
+		return err
+	}
+	if lastEvent != nil {
+		lastEventID = lastEvent.Id
+	}
+
+	// Build on start for orgID 1 but keep lazy for others.
+	_, err = i.getDashboards(ctx, 1)
+	if err != nil {
+		return fmt.Errorf("can't build dashboard search index for org ID 1: %w", err)
+	}
+
+	for {
+		select {
+		case <-partialUpdateTicker.C:
+			lastEventID = i.applyIndexUpdates(context.Background(), lastEventID)
+		case <-fullReIndexTicker.C:
+			started := time.Now()
+			i.reIndexFromScratch()
+			i.logger.Info("Full re-indexing finished", "full_re_index_elapsed", time.Since(started))
```

Comment:
Not sure we have a convention regarding key names, but feels like we're using  camel case is most of our log messages. Maybe use the key `elapsed` for all three log messages since you can differentiate them using the message?

Reply:
Related comment: https://github.com/grafana/grafana/pull/47709#discussion_r851173177 I think we need to have unique names, but I am not sure whether we need to use snake case. cc @FZambia

Reply:
Replaced with camelCased labels. I am worried to make labels non-unique as I don't understand how Loki works and whether we will be able to use https://grafana.com/docs/loki/latest/logql/metric_queries/#unwrapped-range-aggregations effectively or not

Reply:
Make sense :+1: We'll probably add metrics eventually for this if important.